File: typescript/test/nodeMachine.test.ts
import { describe, expect, test } from "vitest";

import { NodeMachine } from "../src/kahnQueue/nodeMachine.js";

describe("NodeMachine", () => {
  test("create_withZeroDependencies_startsReady", () => {
    const m = NodeMachine.create(123, 0);
    expect(m.id).toBe(123);
    expect(m.numSources).toBe(0);
    expect(m.is("READY")).toBe(true);
  });

  test("create_withPositiveDependencies_staysQueuedUntilDecrementedToZero", () => {
    const m = NodeMachine.create(0, 2);
    expect(m.is("QUEUED")).toBe(true);
    expect(m.canTransition("READY")).toBe(false);

    m.decrement();
    expect(m.numSources).toBe(1);
    expect(m.is("QUEUED")).toBe(true);

    m.decrement();
    expect(m.numSources).toBe(0);
    expect(m.is("READY")).toBe(true);
  });

  test("decrement_throwsIfAlreadyZero", () => {
    const m = NodeMachine.create(0, 0);
    expect(() => m.decrement()).toThrow("Attempting to decrement below zero");
  });

  test("transitions_followExpectedLifecycle", () => {
    const m = NodeMachine.create(0, 0);
    expect(m.is("READY")).toBe(true);

    expect(m.canTransition("ACTIVE")).toBe(true);
    m.transition("ACTIVE");
    expect(m.canTransition("COMPLETE")).toBe(true);
    m.transition("COMPLETE");
    expect(m.canTransition("PRUNED")).toBe(false);
  });

  test("pruned_isTerminal_fromQueuedOrReady", () => {
    const queued = NodeMachine.create(0, 1);
    expect(queued.is("QUEUED")).toBe(true);
    queued.transition("PRUNED");
    expect(() => queued.transition("READY")).toThrow(/Invalid transition/);

    const ready = NodeMachine.create(0, 0);
    expect(ready.is("READY")).toBe(true);
    ready.transition("PRUNED");
    expect(() => ready.transition("ACTIVE")).toThrow(/Invalid transition/);
  });
});

File: java/src/test/java/io/github/flashlock/KahnQueue/NodeMachineTest.java
package io.github.flashlock.KahnQueue;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class NodeMachineTest {

  @Test
  void create_withZeroDependencies_startsReady() {
    NodeMachine m = NodeMachine.create(123, 0);
    assertEquals(123, m.id);
    assertEquals(0, m.numSources);
    assertTrue(m.is(NodeState.READY));
  }

  @Test
  void create_withPositiveDependencies_staysQueuedUntilDecrementedToZero() {
    NodeMachine m = NodeMachine.create(0, 2);
    assertTrue(m.is(NodeState.QUEUED));
    assertFalse(m.canTransition(NodeState.READY));

    m.decrement();
    assertEquals(1, m.numSources);
    assertTrue(m.is(NodeState.QUEUED));

    m.decrement();
    assertEquals(0, m.numSources);
    assertTrue(m.is(NodeState.READY));
  }

  @Test
  void decrement_throwsIfAlreadyZero() {
    NodeMachine m = NodeMachine.create(0, 0);
    IllegalStateException ex = assertThrows(IllegalStateException.class, m::decrement);
    assertEquals("Attempting to decrement below zero", ex.getMessage());
  }

  @Test
  void transitions_followExpectedLifecycle() {
    NodeMachine m = NodeMachine.create(0, 0);
    assertTrue(m.is(NodeState.READY));

    assertTrue(m.canTransition(NodeState.ACTIVE));
    m.transition(NodeState.ACTIVE);
    assertTrue(m.canTransition(NodeState.COMPLETE));
    m.transition(NodeState.COMPLETE);
    assertFalse(m.canTransition(NodeState.PRUNED));
  }

  @Test
  void pruned_isTerminal_fromAnyEarlierPhase() {
    NodeMachine queued = NodeMachine.create(0, 1);
    assertTrue(queued.is(NodeState.QUEUED));
    queued.transition(NodeState.PRUNED);
    assertThrows(IllegalStateException.class, () -> queued.transition(NodeState.READY));

    NodeMachine ready = NodeMachine.create(0, 0);
    assertTrue(ready.is(NodeState.READY));
    ready.transition(NodeState.PRUNED);
    assertThrows(IllegalStateException.class, () -> ready.transition(NodeState.ACTIVE));
  }
}

